gprofiler/utils/node.py

```diff
@@ -0,0 +1,223 @@
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def _start_debugger(pid):
+    os.kill(pid, signal.SIGUSR1)
+
+
+def get_node_processes():
+    return pgrep_maps(r"(?:^.+/node[^/]*$)")
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
+    if (
+        debugger_url_response.status_code != 200
+        or not "application/json" in debugger_url_response.headers.get("Content-Type")
+    ):
+        raise NodeDebuggerUrlNotFound
+
+    response_json = debugger_url_response.json()
+    if (
+        not isinstance(response_json, list)
+        or len(response_json) == 0
+        or not isinstance(response_json[0], dict)
+        or not "webSocketDebuggerUrl" in response_json[0].keys()
+    ):
+        raise NodeDebuggerUrlNotFound
+
+    return response_json[0]["webSocketDebuggerUrl"]
+
+
+@retry(NodeDebuggerUnexpectedResponse, 5, 1)
+def _load_dso(sock: WebSocket, module_path: str):
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": f'process.mainModule.require("{os.path.join(module_path, "linux-perf.js")}").start()',
+            "replMode": True,
+        },
+    }
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
+    ):
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+@retry(NodeDebuggerUnexpectedResponse, 5, 1)
+def _stop_dso(sock: WebSocket, module_path: str):
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": f'process.mainModule.require("{os.path.join(module_path, "linux-perf.js")}").stop()',
+            "replMode": True,
+        },
+    }
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
+    ):
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _copy_module_into_process_ns(process: psutil.Process, musl: bool, version: str):
+    proc_root = get_proc_root_path(process)
```

Comment:
Use `resolve_proc_root_links` to resolve the actualy path. If `/tmp` is an absolute symlink in the container, your current code won't work - the path generated in line 124 will eventually point the host. `resolve_proc_root_links` handles that for you

Reply:
The usage now (`resolve_proc_root_links(root, "/")`) will always result in returning `root` because `/` is never a link. But if `dest_inside_container` is a link, the code won't work. Need to call `resolve_proc_root_links(root, dest_inside_container)` and `dest_inside_container` needs to be an absolute path. This way, `resolve_proc_root_links` will follow any symlinks in `dest_inside_container` and correctly re-attach the `/proc/pid/root` part.